### ISSUE TITLE
MLX Phase 1B: NumPy Pruning Accumulator

### ISSUE THREAD

**Issue by egesabanci** [egesabanci/reap-mlx#3]
## Objective

Implement the MLX pruning-metric accumulator using small CPU-side NumPy state while preserving the current REAP observer-data schema used by the Torch pipeline.

## Background

The existing Torch pruning metrics live in `src/reap/pruning_metrics.py`. For MLX, the selected expert outputs and router scores should be computed in MLX, then compact arrays should be transferred to NumPy for accumulation. This avoids relying on missing or awkward MLX scatter/count APIs and avoids retaining lazy MLX graphs across calibration batches.

This issue should produce observer data compatible with pruning methods such as `frequency`, `ean_sum`, `ean_mean`, `weighted_ean_sum`, `weighted_expert_frequency_sum`, `reap`, and `max_activations`.

## Scope

Add only the accumulator and tests. This issue should not implement layer replay or model mutation.

Target files:

- `src/reap/backends/mlx/metrics.py`
- `tests/test_mlx_metrics.py`

## Implementation Steps

1. Create `src/reap/backends/mlx/metrics.py` without importing `torch` or `vllm`.
2. Implement a `PruningState` dataclass with at least:
   - `num_experts`
   - `total_tokens`
   - `expert_frequency`
   - `pairwise_expert_frequency`
   - `ean_sum`
   - `weighted_ean_sum`
   - `weighted_expert_frequency_sum`
   - `max_activations`
3. Add `PruningState.initialize(num_experts)`.
4. Add an accumulation method that accepts routing output plus selected expert outputs or precomputed selected-output norms/maxes.
5. Count `total_tokens` as valid input tokens, not flattened top-k routes.
6. Count `expert_frequency` as routed selections. With `top_k > 1`, `expert_frequency.sum()` may equal `total_tokens * top_k`.
7. Compute `pairwise_expert_frequency` as `freq[:, None] + freq[None, :]` per batch for schema parity.
8. Use `np.bincount` for expert frequency.
9. Use `np.add.at` for per-expert accumulation.
10. Compute derived fields in `report()`:
    - `expert_proba = expert_frequency / max(total_tokens, 1)`
    - `ean_mean = ean_sum / max(expert_frequency, eps)`
    - `reap = weighted_ean_sum / max(expert_frequency, eps)`
11. Include safe behavior for empty batches and never-selected experts.
12. Keep dtypes deliberate:
    - frequency: `int64`
    - sums: `float64`
    - max activations and derived means: `float32` or clearly documented float dtype.

## Acceptance Criteria

- [ ] `PruningState.initialize()` returns correctly shaped zero arrays.
- [ ] `total_tokens` counts input tokens, not top-k routes.
- [ ] `expert_frequency` counts selected expert routes.
- [ ] `pairwise_expert_frequency` is present in `report()`.
- [ ] `expert_proba` is present in `report()` for pruning compatibility.
- [ ] `ean_sum`, `ean_mean`, `weighted_ean_sum`, `weighted_expert_frequency_sum`, `reap`, and `max_activations` are present in `report()`.
- [ ] Empty accumulation is a no-op except for documented token-count behavior.
- [ ] Never-selected experts do not produce NaN or inf values.
- [ ] Tests validate against a hand-computed NumPy reference.
- [ ] Importing the module does not import `torch` or `vllm`.

## Test Plan

Run:

```bash
pytest tests/test_mlx_no_torch_import.py tests/test_mlx_metrics.py -v
```

Test cases to include:

- initialize zeros and dtypes
- single token, single expert
- multiple tokens, multiple experts, `top_k=1`
- multiple tokens, multiple experts, `top_k>1`
- sequential accumulation across multiple batches
- zero-frequency experts
- max activation update over repeated batches
- weighted REAP values against a manual reference

## Non-Goals

- Do not port `OnlineStatsTracker`.
- Do not implement TTM, characteristic activation, router-logit similarity, or merge metrics.
- Do not implement observer replay.
- Do not implement pruning mutation.

## References

- `src/reap/pruning_metrics.py`
- `implementation-plan-draft.md`, P1-003 and GPT-5.5 addendum
- `docs/reap-mlx-implementation-plan-gpt-v0.2.md`, Metric Accumulation

---

## Portfolio Objective Alignment - 2026-05-31

This issue is part of the Apple Silicon MLX REAP backend effort. The production/official REAP experimentation workflow remains the original PyTorch/CUDA implementation; this MLX work must not destabilize or replace that path.

The MLX backend goal is adapter-driven support for MoE weights, not a one-model port. Qwen3-specific work is bootstrap/reference-adapter work. The implementation should preserve REAP pruning semantics and observer-data schema compatibility while adapting to MLX constraints such as lazy execution, no PyTorch hooks, MLX-LM module layouts, and save/reload behavior.

For this issue, the accumulator must preserve REAP pruning semantics and observer-data schema compatibility so MLX experiments can be compared with CUDA REAP results.
